In [1]:
!pip install ipython-sql
!pip install sqlalchemy
!pip install psycopg2

In [1]:
%load_ext sql

In [2]:
from sqlalchemy import create_engine

In [3]:
%sql postgresql://postgres:Kinglord@localhost/moodle

In [4]:
engine = create_engine('postgresql://postgres:Kinglord@localhost/moodle')

In [5]:
%%sql
Select * From mdl_user Limit 5;

 * postgresql://postgres:***@localhost/moodle
5 rows affected.


[(1, 1, 0, 0, 0, '', 'Germany', 'en', 0, 0, 0, 0, '', 'Male', 'DE'),
 (20, 1, 0, 0, 0, '', 'Germany', 'en', 1549745767, 1562935333, 1549745767, 1562935333, '197.215.32.111', 'Male', 'DE'),
 (15, 1, 0, 0, 0, '', 'Germany', 'en', 1549633236, 1563537002, 1563469481, 1563537002, '41.184.239.200', 'Male', 'DE'),
 (36, 1, 0, 0, 0, '', 'Ghana', 'en', 1549641288, 1559227964, 1549651186, 1559227892, '154.160.1.224', 'Male', 'GH'),
 (44, 1, 0, 0, 0, '', 'Ghana', 'en', 1549637001, 1560095090, 1557475553, 1560094908, '129.122.1.14', 'Male', 'GH')]

# Task 1

1.1 Counting the number of tables in a database

In [9]:
%%sql
Select COUNT(*) FROM information_schema.tables

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(643,)]

We have 643 tables in our database but we will explore some which are the following:
mdl_logstore_standard_log,
mdl_context,
mdl_user,
mdl_course,
mdl_modules,
mdl_course_modules,
mdl_course_modules_completion.


1.2
counting records in each table.

In [14]:
#log store records
%%sql
SELECT COUNT(*) FROM mdl_logstore_standard_log

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(417554,)]

In [15]:
#context records
%%sql
SELECT COUNT(*) FROM mdl_context

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(4359,)]

In [16]:
#user recorded
%%sql
SELECT COUNT(*) FROM mdl_user

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(1052,)]

In [17]:
#courses recorded
%%sql
SELECT COUNT(*) FROM mdl_course

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(15,)]

In [18]:
#modules recorded
%%sql
SELECT COUNT(*) FROM mdl_modules

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(26,)]

In [20]:
#course module recorded
%%sql
SELECT COUNT(*) FROM mdl_course_modules

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(290,)]

In [21]:
#modules completed records
%%sql
SELECT COUNT(*) FROM mdl_course_modules_completion

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(4483,)]

In [46]:
%%sql
Select * From mdl_logstore_standard_log Limit 5;

 * postgresql://postgres:***@localhost/moodle
5 rows affected.


[(1, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548311114, 'web', '105.233.48.139', None),
 (2, '\\core\\event\\user_loggedin', 'core', 'loggedin', 'user', 'user', 2, 'r', 0, 1, 10, 0, 2, 0, None, 0, 'a:1:{s:8:"username";s:5:"admin";}', 1548311148, 'web', '105.233.48.139', None),
 (3, '\\core\\event\\dashboard_viewed', 'core', 'viewed', 'dashboard', None, None, 'r', 0, 5, 30, 2, 2, 0, 2, 0, 'N;', 1548311179, 'web', '105.233.48.139', None),
 (4, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548313994, 'web', '105.233.48.139', None),
 (5, '\\core\\event\\course_viewed', 'core', 'viewed', 'course', None, None, 'r', 2, 2, 50, 1, 0, 1, None, 0, 'N;', 1548319150, 'web', '41.13.254.147', None)]

1.3 Number of quiz submissions by hour of day


In [50]:
%%sql
SELECT COUNT(id), EXTRACT(HOUR FROM to_timestamp(timecreated)) from mdl_logstore_standard_log
WHERE action='submitted' AND component='mod_quiz'
GROUP BY EXTRACT(HOUR FROM to_timestamp(timecreated))
ORDER BY count desc;

 * postgresql://postgres:***@localhost/moodle
24 rows affected.


[(104, 13.0),
 (95, 11.0),
 (90, 14.0),
 (88, 12.0),
 (80, 15.0),
 (70, 18.0),
 (69, 10.0),
 (68, 17.0),
 (68, 16.0),
 (57, 19.0),
 (55, 20.0),
 (53, 22.0),
 (46, 9.0),
 (46, 21.0),
 (34, 8.0),
 (30, 1.0),
 (29, 2.0),
 (26, 23.0),
 (23, 0.0),
 (18, 6.0),
 (17, 7.0),
 (14, 4.0),
 (10, 5.0),
 (7, 3.0)]

Many quizes were submitted at 13hr

1.4 Monthly usage time of learners who have confirmed and are not deleted

In [35]:
%%sql
SELECT * FROM mdl_user
WHERE confirmed = '1' AND deleted = '0'
LIMIT 5

 * postgresql://postgres:***@localhost/moodle
5 rows affected.


[(1, 1, 0, 0, 0, '', 'Germany', 'en', 0, 0, 0, 0, '', 'Male', 'DE'),
 (20, 1, 0, 0, 0, '', 'Germany', 'en', 1549745767, 1562935333, 1549745767, 1562935333, '197.215.32.111', 'Male', 'DE'),
 (15, 1, 0, 0, 0, '', 'Germany', 'en', 1549633236, 1563537002, 1563469481, 1563537002, '41.184.239.200', 'Male', 'DE'),
 (36, 1, 0, 0, 0, '', 'Ghana', 'en', 1549641288, 1559227964, 1549651186, 1559227892, '154.160.1.224', 'Male', 'GH'),
 (44, 1, 0, 0, 0, '', 'Ghana', 'en', 1549637001, 1560095090, 1557475553, 1560094908, '129.122.1.14', 'Male', 'GH')]

In [6]:
%%sql 
SELECT COUNT(lastaccess - firstaccess) AS usagetime, EXTRACT (MONTH FROM to_timestamp(firstaccess)) AS month 
FROM mdl_user WHERE confirmed = 1 AND deleted = 0 GROUP BY EXTRACT (MONTH FROM to_timestamp(firstaccess)) 

 * postgresql://postgres:***@localhost/moodle
7 rows affected.


[(58, 3.0),
 (31, 7.0),
 (450, 1.0),
 (227, 2.0),
 (27, 6.0),
 (139, 4.0),
 (112, 5.0)]

1.5 Count of log events per user for the following verbs: ['loggedin', 'viewed', 'started', ,'submitted', 'uploaded', 'updated',
                                                       'searched', 'resumed', 'answered', 'attempted', 'abandoned']

In [36]:
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'viewed'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(120167,)]

In [37]:
#logged in logs
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'loggedin'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(4444,)]

In [38]:
#logs in started status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'started'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(3679,)]

In [39]:
#logs in submitted status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'submitted'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(2288,)]

In [40]:
#logs in uploaded status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'uploaded'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(173,)]

In [41]:
#logs in searched status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'searched'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(14,)]

In [42]:
#logs in resumed status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'resumed'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(1422,)]

In [43]:
#logs in answered status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'answered'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(362,)]

In [44]:
#logs in attempted status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'attempted'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(0,)]

In [45]:
#logs in abandoned status
%%sql
SELECT COUNT(*)
FROM mdl_logstore_standard_log
WHERE action = 'abandoned'

 * postgresql://postgres:***@localhost/moodle
1 rows affected.


[(5,)]

1.6. Write a python class to pull 
Overall grade of learners for 
Number of forum posts


In [12]:
class DbConnection:
    def open_local_db(self,**kwargs):
        # Postgres username, password, and database name
        POSTGRES_ADDRESS = '127.0.0.1' 
        POSTGRES_PORT = '5432'
        POSTGRES_USERNAME = kwargs.get('POSTGRES_USERNAME') ##  YOUR POSTGRES USERNAME
        POSTGRES_PASSWORD =kwargs.get('POSTGRES_PASSWORD')## YOUR POSTGRES PASSWORD
        POSTGRES_DBNAME = kwargs.get('POSTGRES_DBNAME') ## YOUR DATABASE NAME
        # A string that contains the necessary Postgres login information
        postgres_str = ('postgresql://{username}:{password}@{ipaddress}:{port}/{dbname}'.format(username=POSTGRES_USERNAME,password=POSTGRES_PASSWORD,ipaddress=POSTGRES_ADDRESS,port=POSTGRES_PORT,dbname=POSTGRES_DBNAME))
        # Create the connection
        connection = create_engine(postgres_str)
        return connection

In [8]:
kwargs = {'POSTGRES_USERNAME':'postgres', 'POSTGRES_PASSWORD': 'Kinglord', 'POSTGRES_DBNAME':'moodle'}

In [13]:
db = DbConnection()
moodle = db.open_local_db(**kwargs)

In [14]:
import pandas as pd

In [15]:
pd.read_sql_query('''SELECT userid, action,count(action) AS Count FROM mdl_logstore_standard_log 
GROUP BY action ,userid 
ORDER BY count DESC ;''', moodle)

userid    action  count
0          2   created  12175
1        246      sent  10935
2          2  assigned   9493
3        581      sent   9087
4          2      sent   8025
...      ...       ...    ...
6920     532     shown      1
6921    1001  loggedin      1
6922      84     reset      1
6923     611   created      1
6924     448  loggedin      1

[6925 rows x 3 columns]